core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1671,143 +1671,157 @@ final class NamesResolved implements Runnable {
         @SuppressWarnings("ReferenceEquality")
         @Override
         public void run() {
-          if (ManagedChannelImpl.this.nameResolver != resolver) {
-            return;
-          }
+          onResult2(resolutionResult, true);
+        }
+      }
 
-          List<EquivalentAddressGroup> servers = resolutionResult.getAddresses();
-          channelLogger.log(
-              ChannelLogLevel.DEBUG,
-              "Resolved address: {0}, config={1}",
-              servers,
-              resolutionResult.getAttributes());
-
-          if (lastResolutionState != ResolutionState.SUCCESS) {
-            channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}", servers);
-            lastResolutionState = ResolutionState.SUCCESS;
-          }
+      syncContext.execute(new NamesResolved());
+    }
 
-          ConfigOrError configOrError = resolutionResult.getServiceConfig();
-          ResolutionResultListener resolutionResultListener = resolutionResult.getAttributes()
-              .get(RetryingNameResolver.RESOLUTION_RESULT_LISTENER_KEY);
-          InternalConfigSelector resolvedConfigSelector =
-              resolutionResult.getAttributes().get(InternalConfigSelector.KEY);
-          ManagedChannelServiceConfig validServiceConfig =
-              configOrError != null && configOrError.getConfig() != null
-                  ? (ManagedChannelServiceConfig) configOrError.getConfig()
-                  : null;
-          Status serviceConfigError = configOrError != null ? configOrError.getError() : null;
-
-          ManagedChannelServiceConfig effectiveServiceConfig;
-          if (!lookUpServiceConfig) {
-            if (validServiceConfig != null) {
-              channelLogger.log(
-                  ChannelLogLevel.INFO,
-                  "Service config from name resolver discarded by channel settings");
-            }
-            effectiveServiceConfig =
-                defaultServiceConfig == null ? EMPTY_SERVICE_CONFIG : defaultServiceConfig;
-            if (resolvedConfigSelector != null) {
+    @Override
+    public Status onResult2(final ResolutionResult resolutionResult) {
+      return onResult2(resolutionResult, false);
+    }
+
+    private Status onResult2(final ResolutionResult resolutionResult, boolean useResolutionResultListener) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      if (ManagedChannelImpl.this.nameResolver != resolver) {
+        return Status.FAILED_PRECONDITION;
```

Comment:
As a rule, don't use bare status codes for errors. It is very difficult to find where they came from. Always include at least a description or cause.

Although I think in both of these FAILED_PRECONDITION cases we could just return OK instead. We aren't rejecting the config so much all future updates will be rejected; there's no need for the name resolver to retry.

Reply:
Since the check 


if (ManagedChannelImpl.this.nameResolver != resolver) 

 is already made here, do we still need the check

if (NameResolverListener.this.helper == ManagedChannelImpl.this.lbHelper) {

later in the same method, which can never happen since the method runs inside the syncContext?

I added a new unit test noMoreCallbackAfterLoadBalancerShutdown_configError to test that the retry for name resolving doesn't happen after the onResult has a config error. However I'm finding that the test succeeds even if I don't change returning the FAILED_PRECONDITION to OK, and I notice that the scheduling for next retry here happens https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/RetryingNameResolver.java#L107 with 0.9 s delay but it doesn't get triggered even after adding a sleep statement longer than that in the test. Still looking into this.

Reply:
I realized the executor is a fake executor service, so I have added the assertion on the fake clock pending task now to ensure that no retry of the name resolution gets scheduled for the case of config error in ResolutionResult.